diff -urN a/arch/arm/boot/dts/ev-imx6ul-m2.dts b/arch/arm/boot/dts/ev-imx6ul-m2.dts
--- a/arch/arm/boot/dts/ev-imx6ul-m2.dts	1970-01-01 03:00:00.000000000 +0300
+++ b/arch/arm/boot/dts/ev-imx6ul-m2.dts	2020-07-03 17:08:26.471831119 +0300
@@ -0,0 +1,530 @@
+/*
+ * Copyright (C) 2015 Freescale Semiconductor, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ */
+
+ /*
+  * EV-iMX6UL-NANO module
+  * www.evodbg.com
+  * contact  r.sariev@evodbg.com
+  */ 
+
+/dts-v1/;
+
+#include <dt-bindings/input/input.h>
+#include "imx6ull.dtsi"
+
+
+/ {
+	model = "Evodbg i.MX6ULL M2 format module";
+	compatible = "fsl,imx6ull-14x14-evk", "fsl,imx6ull";
+
+	chosen {
+		stdout-path = &uart1;
+	};
+
+	memory {
+		reg = <0x80000000 0x20000000>;
+	};
+
+	reserved-memory {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		ranges;
+
+		linux,cma {
+			compatible = "shared-dma-pool";
+			reusable;
+			size = <0x14000000>;
+			linux,cma-default;
+		};
+	};
+
+	regulators {
+		compatible = "simple-bus";
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		reg_can_3v3: regulator@0 {
+			compatible = "regulator-fixed";
+			reg = <0>;
+			regulator-name = "can-3v3";
+			regulator-min-microvolt = <3300000>;
+			regulator-max-microvolt = <3300000>;
+			regulator-always-on;
+		};
+
+		reg_sd1_vmmc: regulator@1 {
+			compatible = "regulator-fixed";
+			regulator-name = "VSD_3V3";
+			regulator-min-microvolt = <3300000>;
+			regulator-max-microvolt = <3300000>;
+			regulator-always-on;
+		};
+
+		reg_gpio_dvfs: regulator-gpio {
+                        compatible = "regulator-fixed";
+                        regulator-min-microvolt = <1400000>;
+                        regulator-max-microvolt = <1400000>;
+                        regulator-name = "gpio_dvfs";
+                        regulator-always-on;
+                };
+
+		
+                reg_usb1_vbus: regulator@2 {
+                        compatible = "regulator-fixed";
+                        reg = <2>;
+                        regulator-name = "usb1_vbus";
+                        regulator-min-microvolt = <5000000>;
+                        regulator-max-microvolt = <5000000>;
+			regulator-always-on;
+                };
+
+
+		reg_usb2_vbus: regulator@3 {
+			reg = <3>;
+			compatible = "regulator-fixed";
+			regulator-name = "usb2_vbus";
+			regulator-min-microvolt = <5000000>;
+			regulator-max-microvolt = <5000000>;
+			regulator-always-on;
+		};
+		reg_lcd_fbtft: regulator@4 {
+			reg = <4>;
+			regulator-name = "lcd_3v3";
+                        regulator-min-microvolt = <3300000>;
+                        regulator-max-microvolt = <3300000>;
+			regulator-always-on;
+                };
+	};
+		sound {
+	};
+};
+
+&cpu0 {
+        arm-supply = <&reg_arm>;
+        soc-supply = <&reg_soc>;
+        dc-supply = <&reg_gpio_dvfs>;
+};
+
+
+&clks {
+	assigned-clocks = <&clks IMX6UL_CLK_PLL4_AUDIO_DIV>;
+	assigned-clock-rates = <786432000>;
+};
+
+
+&uart1 {					// DUART console / UART1
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_uart1>;
+	status = "okay";
+};
+
+
+&uart2 { 					// UART2 RS485 mode
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_uart2>;
+	fsl,uart-has-rtscts;
+	linux,rs485-enabled-at-boot-time;
+	status = "okay";
+};
+
+
+&uart3 {					// UART3 RS485 mode
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_uart3>;
+	fsl,uart-has-rtscts;
+	linux,rs485-enabled-at-boot-time;
+	status = "okay";
+};
+
+
+&uart4 {					// UART4 
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_uart4>;
+	status = "okay";
+};
+
+
+&ecspi1 {
+	fsl,spi-num-chipselects = <1>;
+        cs-gpios = <&gpio3 26 0>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_ecspi1>;
+	status = "okay";
+
+	spidev@0 {
+		rotate = <90>;
+                bgr;
+                fps = <30>;
+                compatible = "ilitek,ili9341";
+                spi-max-frequency = <50000000>;   
+                reg = <0>;
+                regwidth = <8>;
+                buswidth = <8>;
+                verbose = <3>;
+		dc-gpios = <&gpio1 11 GPIO_ACTIVE_LOW>;    /* JTAG_TMS */
+		vbus-supply = <&reg_lcd_fbtft>;
+
+	};
+};
+
+&fec2 {
+        pinctrl-names = "default";
+        pinctrl-0 = <&pinctrl_enet2>;
+        phy-mode = "rmii";
+        phy-reset-gpios = <&gpio3 23 1>;
+        phy-reset-duration = <250>;
+        phy-handle = <&ethphy1>;
+        status = "okay";
+
+        mdio {
+                #address-cells = <1>;
+                #size-cells = <0>;
+
+                ethphy1: ethernet-phy@1 {
+                        compatible = "ethernet-phy-ieee802.3-c22";
+                        reg = <0>;
+                };
+        };
+};
+
+
+
+/* flexcan2 muxed with UART2 RTS/CTS pins */
+&can2 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_flexcan2>;
+	xceiver-supply = <&reg_can_3v3>;
+	status = "okay";
+};
+
+&gpc {
+	fsl,cpu_pupscr_sw2iso = <0x1>;
+	fsl,cpu_pupscr_sw = <0x0>;
+	fsl,cpu_pdnscr_iso2sw = <0x1>;
+	fsl,cpu_pdnscr_iso = <0x1>;
+	fsl,ldo-bypass = <0>; /* DCDC, ldo-enable */
+};
+
+
+&i2c2 {
+	clock_frequency = <100000>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_i2c2>;
+	status = "okay";
+};
+
+&iomuxc_snvs {
+        pinctrl-names = "default_snvs";
+        pinctrl-0 = <&pinctrl_hog_2>;
+        imx6ul-evk {
+                pinctrl_hog_2: hoggrp-2 {
+                        fsl,pins = <
+//				MX6UL_PAD_LCD_DATA18__GPIO3_IO23	0x79
+                        >;
+		};
+	};
+};
+
+
+
+&iomuxc {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_hog_1>;
+	imx6ul-evk {
+		pinctrl_hog_1: hoggrp-1 {
+			fsl,pins = <
+				MX6UL_PAD_LCD_DATA18__GPIO3_IO23	0x1b0b0
+				MX6UL_PAD_UART1_RTS_B__GPIO1_IO19       0x17059    /* SD1 CD */
+//				MX6UL_PAD_CSI_DATA04__GPIO4_IO25	0x1b0b0	   /* GPIO4_25 - gpio121 */
+//				MX6UL_PAD_CSI_DATA05__GPIO4_IO26	0x1b0b0    /* GPIO4_26 - gpio122 */
+//				MX6UL_PAD_CSI_DATA06__GPIO4_IO27	0x1b0b0    /* GPIO4_27 - gpio123 */
+//				MX6UL_PAD_CSI_DATA07__GPIO4_IO28	0x1b0b0    /* GPIO4_28 - gpio124  */
+//				MX6UL_PAD_UART2_RTS_B__GPIO1_IO23	0x1b0b0	   /* UART2-RTS - LCD DC */
+//				MX6UL_PAD_UART3_RTS_B__GPIO1_IO27	0x1b0b0	   /* UART3-RTS - LCD RESET */
+
+
+// Main Connector                                                                     Module pin
+//				MX6UL_PAD_CSI_MCLK__GPIO4_IO17				/* 21 / CSI_MCLK */
+//				MX6UL_PAD_CSI_PIXCLK__GPIO4_IO18			/* 23 / CSI_PIXCLK */
+//				MX6UL_PAD_CSI_VSYNC__GPIO4_IO19				/* 25 / CSI_VSYNC */
+//				MX6UL_PAD_UART4_TX_DATA__GPIO1_IO28			/* 26 / UART4_TX
+//				MX6UL_PAD_CSI_HSYNC__GPIO4_IO20				/* 27 / CSI_HSYNC */
+//				MX6UL_PAD_UART4_RX_DATA__GPIO1_IO29			/* 28 / UART4_RX */
+//				MX6UL_PAD_CSI_DATA00__GPIO4_IO21			/* 29 / CSI_DATA00 */
+//				MX6UL_PAD_UART5_TX_DATA__GPIO1_IO30			/* 30 / I2C2_SCL */
+//			    	MX6UL_PAD_CSI_DATA01__GPIO4_IO22			/* 31 / CSI_DATA01 */
+//				MX6UL_PAD_UART5_RX_DATA__GPIO1_IO31			/* 32 / I2C2_SDA */
+//				MX6UL_PAD_CSI_DATA02__GPIO4_IO23			/* 33 / CSI_DATA02 */
+//				MX6UL_PAD_LCD_DATA20__GPIO3_IO25			/* 34 / ECSPI1_SCLK */
+//				MX6UL_PAD_CSI_DATA03__GPIO4_IO24			/* 35 / CSI_DATA03 */
+//				MX6UL_PAD_LCD_DATA23__GPIO3_IO28			/* 36 / ECSPI1_MISO */
+//				MX6UL_PAD_CSI_DATA04__GPIO4_IO25			/* 37 / CSI_DATA04 */
+//				MX6UL_PAD_LCD_DATA22__GPIO3_IO27			/* 38 / ECSPI1_MOSI */
+//				MX6UL_PAD_CSI_DATA05__GPIO4_IO26			/* 39 / CSI_DATA05 */
+//				MX6UL_PAD_LCD_DATA21__GPIO3_IO26			/* 40 / ECSPI1_SS0 */
+//				MX6UL_PAD_CSI_DATA06__GPIO4_IO27			/* 41 / CSI_DATA06 */
+//				MX6UL_PAD_LCD_DATA05__GPIO3_IO10			/* 42 / ECSPI1_SS1 */
+//				MX6UL_PAD_CSI_DATA07__GPIO4_IO28			/* 43 / CSI_DATA07 */
+//				MX6UL_PAD_UART1_TX_DATA__GPIO1_IO16			/* 45 / UART1_TX */
+//				MX6UL_PAD_UART1_RX_DATA__GPIO1_IO17			/* 47 / UART1_RX */
+//				MX6UL_PAD_GPIO1_IO04__GPIO1_IO04			/* 48 / PWM3 */
+//				MX6UL_PAD_JTAG_MOD__GPIO1_IO10				/* 49 / SPDIF_OUT */
+//				MX6UL_PAD_JTAG_TCK__GPIO1_IO14				/* 50 / SAI2_RXD */
+//				MX6UL_PAD_SD1_CLK__GPIO2_IO17				/* 51 / SD1_CLK */
+//				MX6UL_PAD_JTAG_TMS__GPIO1_IO11		0x1b0b0		/* 52 / SAI2_MCLK */
+//				MX6UL_PAD_SD1_CMD__GPIO2_IO16				/* 53 / SD1_CMD */
+//				MX6UL_PAD_JTAG_TDI__GPIO1_IO13				/* 54 / SAI2_BCLK */
+//				MX6UL_PAD_SD1_DATA0__GPIO2_IO18				/* 55 / SD1_DATA0 */
+//				MX6UL_PAD_JTAG_TDO__GPIO1_IO12				/* 56 / SAI2_SYNC */
+//				MX6UL_PAD_SD1_DATA1__GPIO2_IO19				/* 57 / SD1_DATA1 */
+//				MX6UL_PAD_JTAG_TRST_B__GPIO1_IO15			/* 58 / SAI2_TXD */
+//				MX6UL_PAD_SD1_DATA2__GPIO2_IO20				/* 59 / SD1_DATA2 */
+//				MX6UL_PAD_UART3_TX_DATA__GPIO1_IO24			/* 60 / UART3_TX */
+//				MX6UL_PAD_SD1_DATA3__GPIO2_IO21				/* 61 / SD1_DATA3 */
+//				MX6UL_PAD_UART3_RX_DATA__GPIO1_IO25			/* 62 / UART3_RX */
+//				MX6UL_PAD_UART1_RTS_B__GPIO1_IO19			/* 63 / SD1_CD */
+//				MX6UL_PAD_UART3_RTS_B__GPIO1_IO27			/* 64 / UART3_RTS */
+//				MX6UL_PAD_UART3_CTS_B__GPIO1_IO26			/* 66 / UART3_CTS */
+//				MX6UL_PAD_UART2_TX_DATA__GPIO1_IO20			/* 68 / UART2_TX */
+//				MX6UL_PAD_GPIO1_IO00__GPIO1_IO00			/* 69 / USB1_ID */
+//				MX6UL_PAD_UART2_RX_DATA__GPIO1_IO21			/* 70 / UART2_RX */
+//				MX6UL_PAD_UART2_RTS_B__GPIO1_IO23			/* 72 / UART2_RTS */
+//				MX6UL_PAD_UART2_CTS_B__GPIO1_IO22			/* 74 / UART2_CTS */
+// Full pin function list see in imx6ul-pinfunc.h and imx6ull-pinfunc.h.
+
+
+
+			>;
+		};
+
+		pinctrl_enet2: enet2grp {
+			fsl,pins = <
+				MX6UL_PAD_ENET1_TX_EN__ENET2_MDC	0x1b0b0
+				MX6UL_PAD_GPIO1_IO06__ENET2_MDIO	0x1b0b0
+				MX6UL_PAD_ENET2_RX_EN__ENET2_RX_EN	0x1b0b0
+				MX6UL_PAD_ENET2_RX_ER__ENET2_RX_ER	0x1b0b0
+				MX6UL_PAD_ENET2_RX_DATA0__ENET2_RDATA00	0x1b0b0
+				MX6UL_PAD_ENET2_RX_DATA1__ENET2_RDATA01	0x1b0b0
+				MX6UL_PAD_ENET2_TX_EN__ENET2_TX_EN	0x1b0b0
+				MX6UL_PAD_ENET2_TX_DATA0__ENET2_TDATA00	0x1b0b0
+				MX6UL_PAD_ENET2_TX_DATA1__ENET2_TDATA01	0x1b0b0
+				MX6UL_PAD_ENET2_TX_CLK__ENET2_REF_CLK2	0x4001b031
+			>;
+		};
+
+		pinctrl_flexcan1: flexcan1grp{
+			fsl,pins = <
+//				MX6UL_PAD_UART3_RTS_B__FLEXCAN2_RX	0x1b020
+//				MX6UL_PAD_UART3_CTS_B__FLEXCAN2_TX	0x1b020
+			>;
+		};
+
+
+		pinctrl_flexcan2: flexcan2grp{
+			fsl,pins = <
+//				MX6UL_PAD_UART2_RTS_B__FLEXCAN2_RX	0x1b020
+//				MX6UL_PAD_UART2_CTS_B__FLEXCAN2_TX	0x1b020
+			>;
+		};
+
+		pinctrl_i2c2: i2c2grp {
+                        fsl,pins = <
+                                MX6UL_PAD_UART5_TX_DATA__I2C2_SCL 0x4001b8b0
+                                MX6UL_PAD_UART5_RX_DATA__I2C2_SDA 0x4001b8b0
+                        >;
+                };
+
+		pinctrl_pwm3: pwm3grp {
+			fsl,pins = <
+				MX6UL_PAD_GPIO1_IO04__PWM3_OUT   	0x110b0   	/* ok */
+			>;
+		};
+
+		pinctrl_sai2: sai2grp {
+			fsl,pins = <
+				MX6UL_PAD_JTAG_TDI__SAI2_TX_BCLK        0x17088		/* ok */
+                                MX6UL_PAD_JTAG_TDO__SAI2_TX_SYNC        0x17088
+                                MX6UL_PAD_JTAG_TRST_B__SAI2_TX_DATA     0x11088
+                                MX6UL_PAD_JTAG_TCK__SAI2_RX_DATA        0x11088
+                                MX6UL_PAD_JTAG_TMS__SAI2_MCLK           0x17088
+
+			>;
+		};
+
+		pinctrl_ecspi1: spi1grp {
+			fsl,pins = <
+				MX6UL_PAD_LCD_DATA20__ECSPI1_SCLK	0x1b0b1		/* ok */
+//				MX6UL_PAD_LCD_DATA21__ECSPI1_SS0	0x1b0b1
+				MX6UL_PAD_LCD_DATA21__GPIO3_IO26	0x1b0b1		/* CS0 */
+				MX6UL_PAD_LCD_DATA05__GPIO3_IO10	0x1b0b1		/* CS1 */
+				MX6UL_PAD_LCD_DATA22__ECSPI1_MOSI	0x1b0b1
+				MX6UL_PAD_LCD_DATA23__ECSPI1_MISO	0x1b0b1
+
+			>;
+		};
+
+		pinctrl_uart1: uart1grp {
+			fsl,pins = <
+				MX6UL_PAD_UART1_TX_DATA__UART1_DCE_TX 0x1b0b1		/* ok */
+				MX6UL_PAD_UART1_RX_DATA__UART1_DCE_RX 0x1b0b15
+			>;
+		};
+
+		pinctrl_uart2: uart2grp {
+			fsl,pins = <
+				MX6UL_PAD_UART2_TX_DATA__UART2_DCE_TX	0x1b0b1		/* ok */
+				MX6UL_PAD_UART2_RX_DATA__UART2_DCE_RX	0x1b0b1
+				MX6UL_PAD_UART2_CTS_B__UART2_DCE_CTS    0x1b0b1
+				MX6UL_PAD_UART2_RTS_B__UART2_DCE_RTS	0x1b0b1
+			>;
+		};
+
+		pinctrl_uart3: uart3grp {
+			fsl,pins = <
+				MX6UL_PAD_UART3_TX_DATA__UART3_DCE_TX	0x1b0b1		/* ok */
+				MX6UL_PAD_UART3_RX_DATA__UART3_DCE_RX	0x1b0b1
+				MX6UL_PAD_UART3_CTS_B__UART3_DCE_CTS	0x1b0b1
+				MX6UL_PAD_UART3_RTS_B__UART3_DCE_RTS	0x1b0b1
+			>;
+		};
+
+		pinctrl_uart4: uart4grp {
+			fsl,pins = <
+				MX6UL_PAD_UART4_TX_DATA__UART4_DCE_TX	0x1b0b1		/* ok */
+				MX6UL_PAD_UART4_RX_DATA__UART4_DCE_RX	0x1b0b1
+			>;
+		};
+
+		pinctrl_uart5: uart5grp {
+			fsl,pins = <
+				MX6UL_PAD_CSI_DATA00__UART5_DCE_TX	0x1b0b1
+				MX6UL_PAD_CSI_DATA01__UART5_DCE_RX	0x1b0b1
+				MX6UL_PAD_CSI_DATA02__UART5_DCE_RTS	0x1b0b1
+				MX6UL_PAD_CSI_DATA03__UART5_DCE_CTS	0x1b0b1
+			>;
+		};
+
+		pinctrl_uart6: uart6grp {
+			fsl,pins = <
+				MX6UL_PAD_CSI_MCLK__UART6_DCE_TX	0x1b0b1
+				MX6UL_PAD_CSI_PIXCLK__UART6_DCE_RX	0x1b0b1
+				MX6UL_PAD_CSI_VSYNC__UART6_DCE_RTS	0x1b0b1
+				MX6UL_PAD_CSI_HSYNC__UART6_DCE_CTS	0x1b0b1
+			>;
+		};
+
+		pinctrl_usdhc1: usdhc1grp {
+			fsl,pins = <
+				MX6UL_PAD_SD1_CMD__USDHC1_CMD     0x17059
+				MX6UL_PAD_SD1_CLK__USDHC1_CLK     0x10071
+				MX6UL_PAD_SD1_DATA0__USDHC1_DATA0 0x17059
+				MX6UL_PAD_SD1_DATA1__USDHC1_DATA1 0x17059
+				MX6UL_PAD_SD1_DATA2__USDHC1_DATA2 0x17059
+				MX6UL_PAD_SD1_DATA3__USDHC1_DATA3 0x17059
+			>;
+		};
+
+		pinctrl_usdhc2: usdhc2grp {
+			fsl,pins = <
+				MX6UL_PAD_NAND_RE_B__USDHC2_CLK     0x10069
+				MX6UL_PAD_NAND_WE_B__USDHC2_CMD     0x17059
+				MX6UL_PAD_NAND_DATA00__USDHC2_DATA0 0x17059
+				MX6UL_PAD_NAND_DATA01__USDHC2_DATA1 0x17059
+				MX6UL_PAD_NAND_DATA02__USDHC2_DATA2 0x17059
+				MX6UL_PAD_NAND_DATA03__USDHC2_DATA3 0x17059
+				MX6UL_PAD_NAND_DATA04__USDHC2_DATA4 0x17059
+				MX6UL_PAD_NAND_DATA05__USDHC2_DATA5 0x17059
+				MX6UL_PAD_NAND_DATA06__USDHC2_DATA6 0x17059
+				MX6UL_PAD_NAND_DATA07__USDHC2_DATA7 0x17059
+			>;
+		};
+
+		pinctrl_wdog: wdoggrp {
+			fsl,pins = <
+				MX6UL_PAD_LCD_RESET__WDOG1_WDOG_ANY    0x30b0
+			>;
+		};
+
+//		pinctrl_usb1vbus_gpio: usb1vbusgrp {
+//			fsl,pins = <
+//				MX6UL_PAD_LCD_DATA05__GPIO3_IO10	0x30b0
+//	>;
+//    };
+	};
+};
+
+
+&pwm3 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_pwm3>;
+	status = "okay";
+};
+
+&sai2 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_sai2>;
+
+	assigned-clocks = <&clks IMX6UL_CLK_SAI2_SEL>,
+			  <&clks IMX6UL_CLK_SAI2>;
+	assigned-clock-parents = <&clks IMX6UL_CLK_PLL4_AUDIO_DIV>;
+	assigned-clock-rates = <0>, <12288000>;
+	status = "okay";
+};
+
+&usbotg1 {
+	dr_mode = "host";
+	srp-disable;
+	hnp-disable;
+	adp-disable;
+	vbus-supply = <&reg_usb1_vbus>;
+	status = "okay";
+};
+
+&usbotg2 {
+	dr_mode = "host";
+	disable-over-current;
+	vbus-supply = <&reg_usb2_vbus>;
+	status = "okay";
+};
+
+&usbphy1 {
+	tx-d-cal = <0x5>;
+};
+
+&usbphy2 {
+	tx-d-cal = <0x5>;
+};
+
+&usdhc1 {
+        pinctrl-names = "default";
+        pinctrl-0 = <&pinctrl_usdhc1>;
+        cd-gpios = <&gpio1 19 GPIO_ACTIVE_LOW>;
+        keep-power-in-suspend;
+        enable-sdio-wakeup;
+	no-1-8-v;
+	bus-width = <4>; 
+        vmmc-supply = <&reg_sd1_vmmc>;
+        status = "okay";
+};
+
+&usdhc2 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_usdhc2>;
+	no-1-8-v;
+	bus-width = <8>;
+	non-removable;
+	keep-power-in-suspend;
+	enable-sdio-wakeup;
+	status = "okay";
+};
+
+&wdog1 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_wdog>;
+	fsl,wdog_b;
+};
diff -urN a/arch/arm/boot/dts/ev-o4-imx6ul-nano.dts b/arch/arm/boot/dts/ev-o4-imx6ul-nano.dts
--- a/arch/arm/boot/dts/ev-o4-imx6ul-nano.dts	1970-01-01 03:00:00.000000000 +0300
+++ b/arch/arm/boot/dts/ev-o4-imx6ul-nano.dts	2020-07-03 14:58:09.012652238 +0300
@@ -0,0 +1,561 @@
+ /*
+ * Copyright (C) 2015 Freescale Semiconductor, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ */
+
+ /*
+  * EV-iMX6UL-NANO module
+  * www.evodbg.com
+  * contact  r.sariev@evodbg.com
+  */ 
+
+/dts-v1/;
+
+#include <dt-bindings/input/input.h>
+#include "imx6ull.dtsi"
+
+#define DUAL_ETH   // Comment this line for modules with one Ethernet PHY chip
+
+/ {
+	model = "Evodbg i.MX6ULL module";
+	compatible = "fsl,imx6ull-14x14-evk", "fsl,imx6ull";
+
+	chosen {
+		stdout-path = &uart1;
+	};
+
+	memory {
+		reg = <0x80000000 0x20000000>;
+	};
+
+	reserved-memory {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		ranges;
+
+		linux,cma {
+			compatible = "shared-dma-pool";
+			reusable;
+			size = <0x14000000>;
+			linux,cma-default;
+		};
+	};
+
+	regulators {
+		compatible = "simple-bus";
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		reg_can_3v3: regulator@0 {
+			compatible = "regulator-fixed";
+			reg = <0>;
+			regulator-name = "can-3v3";
+			regulator-min-microvolt = <3300000>;
+			regulator-max-microvolt = <3300000>;
+			regulator-always-on;
+		};
+
+		reg_sd1_vmmc: regulator@1 {
+			compatible = "regulator-fixed";
+			regulator-name = "VSD_3V3";
+			regulator-min-microvolt = <3300000>;
+			regulator-max-microvolt = <3300000>;
+			regulator-always-on;
+		};
+		
+		reg_gpio_dvfs: regulator-gpio {
+                        compatible = "regulator-fixed";
+                        regulator-min-microvolt = <1400000>;
+                        regulator-max-microvolt = <1400000>;
+                        regulator-name = "gpio_dvfs";
+                        regulator-always-on;
+                };
+
+		reg_usb1_vbus: regulator@2 {
+                        compatible = "regulator-fixed";
+                        reg = <2>;
+                        regulator-name = "usb1_vbus";
+                        regulator-min-microvolt = <5000000>;
+                        regulator-max-microvolt = <5000000>;
+                        gpio = <&pcf8574 6 GPIO_ACTIVE_HIGH>;
+                        enable-active-high;
+                        regulator-boot-on;
+                };
+
+
+		reg_usb2_vbus: regulator@3 {
+                        compatible = "regulator-fixed";
+                        reg = <2>;
+                        regulator-name = "usb2_vbus";
+                        regulator-min-microvolt = <5000000>;
+                        regulator-max-microvolt = <5000000>;
+                        gpio = <&pcf8574 7 GPIO_ACTIVE_HIGH>;
+                        enable-active-high;
+                        regulator-boot-on;
+                };
+	};
+		leds    {
+                        compatible = "gpio-leds";
+                    led1 {
+                        gpios = <&pcf8574 0 GPIO_ACTIVE_LOW>;
+                        label = "LED1_RED";
+                        default-state = "on";
+                        linux,default-trigger = "heartbeat";
+                        };
+                    led2 {
+                        gpios = <&pcf8574 1 GPIO_ACTIVE_LOW>;
+                        label = "LED1_GREEN";
+                        linux,default-trigger = "mmc1";
+                        default-state = "off";
+                        };
+                    led3 {
+                        gpios = <&pcf8574 2 GPIO_ACTIVE_LOW>;
+                        label = "LED1_BLUE";
+                        linux,default-trigger = "cpu0";
+                        default-state = "on";
+                        };
+                    led4 {
+                        gpios = <&pcf8574 3 GPIO_ACTIVE_LOW>;
+                        label = "LED2_RED";
+                        linux,default-trigger = "mmc0";
+                        default-state = "on";
+                        };
+		    led5 {
+                        gpios = <&pcf8574 4 GPIO_ACTIVE_LOW>;
+                        label = "LED2_GREEN";
+                        default-state = "off";
+                        };
+		    led6 {
+                        gpios = <&pcf8574 5 GPIO_ACTIVE_LOW>;
+                        label = "LED2_BLUE";
+                        default-state = "off";
+                        };
+        };
+
+};
+
+&cpu0 {
+	arm-supply = <&reg_arm>;
+	soc-supply = <&reg_soc>;
+	dc-supply = <&reg_gpio_dvfs>;
+};
+
+&clks {
+	assigned-clocks = <&clks IMX6UL_CLK_PLL4_AUDIO_DIV>;
+	assigned-clock-rates = <786432000>;
+};
+/*
+&csi {
+	status = "disabled";
+};
+*/
+
+&uart1 {					// DUAURT console
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_uart1>;
+	status = "okay";
+};
+
+
+&uart2 { 					// UART2 RS485
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_uart2>;
+	uart-has-rtscts;
+	rts-gpios = <&gpio1 27  GPIO_ACTIVE_HIGH>;// UART3_RTS (32 Module pin) 
+	linux,rs485-enabled-at-boot-time;
+	status = "okay";
+};
+
+
+&uart3 {					// UART3
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_uart3>;
+	status = "okay";
+};
+
+
+&uart4 {					// UART4
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_uart4>;
+	status = "okay";
+};
+
+
+&ecspi1 {
+        fsl,spi-num-chipselects = <1>;
+        cs-gpios = <&gpio4 26 0>;
+        pinctrl-names = "default";
+        pinctrl-0 = <&pinctrl_ecspi1>;
+        status = "okay";
+
+        ili9341@0 {
+                rotate = <90>;
+                bgr;
+                fps = <30>;
+                compatible = "ilitek,ili9341";
+                spi-max-frequency = <50000000>;
+                reg = <0>;
+                regwidth = <8>;
+                buswidth = <8>;
+                verbose = <3>;
+                dc-gpios = <&gpio4 22 GPIO_ACTIVE_LOW>;   
+                vbus-supply = <&reg_can_3v3>;
+
+        };
+};
+
+
+&fec1 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_enet1>;
+	phy-mode = "rmii";
+	phy-handle = <&ethphy0>;
+	status = "okay";
+};
+
+
+
+&fec2 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_enet2>;
+	phy-mode = "rmii";
+	phy-reset-gpios = <&gpio5 1 1>;
+	phy-reset-duration = <250>;
+	phy-handle = <&ethphy1>;
+	status = "okay";
+
+	mdio {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		ethphy0: ethernet-phy@0 {
+			compatible = "ethernet-phy-ieee802.3-c22";
+			reg = <0>;
+		};
+
+		ethphy1: ethernet-phy@1 {
+			compatible = "ethernet-phy-ieee802.3-c22";
+			reg = <1>;
+		};
+	};
+};
+
+
+
+&can1 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_flexcan1>;
+	xceiver-supply = <&reg_can_3v3>;
+	status = "okay";
+};
+
+
+
+&gpc {
+	fsl,cpu_pupscr_sw2iso = <0x1>;
+	fsl,cpu_pupscr_sw = <0x0>;
+	fsl,cpu_pdnscr_iso2sw = <0x1>;
+	fsl,cpu_pdnscr_iso = <0x1>;
+	fsl,ldo-bypass = <0>; /* DCDC, ldo-enable */
+};
+
+&i2c2 {
+        clock_frequency = <100000>;
+        pinctrl-names = "default";
+        pinctrl-0 = <&pinctrl_i2c2>;
+        status = "okay";
+                pcf8574: gpio@38 {
+                                        //GPIO Expander1
+                                            // P0 - LED1-RED
+                                            // P1 - LED1-GREEN
+                                            // P2 - LED1-BLUE
+                                            // P3 - LED2-RED
+                                            // P4 - LED2-GREEN
+                                            // P5 - LED2-BLUE
+                                            // P6 - USB0-PWR
+                                            // P7 - USB1-PWR
+                        compatible = "nxp,pcf8574a";
+                        reg = <0x38>;       // PCF8574A Address -0-0-0
+                        gpio-controller;
+                        #gpio-cells = <2>;
+        };
+};
+
+
+
+
+
+&iomuxc {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_hog_1>;
+	imx6ul-evk {
+		pinctrl_hog_1: hoggrp-1 {
+			fsl,pins = <
+
+				MX6UL_PAD_UART1_RTS_B__GPIO1_IO19       0x17059    	/* SD1 CD 			*/
+//				MX6UL_PAD_UART3_RTS_B__GPIO1_IO27	0x1b0b0		/* GPIO 27 	UART2_DIR	*/
+//				MX6UL_PAD_CSI_DATA02__GPIO4_IO23	0x1b0b0		/* GPIO 119 	UART3_DIR	*/
+//				MX6UL_PAD_LCD_DATA05__GPIO3_IO10	0x1b0b0		/* GPIO 74			*/
+
+
+// Main Connector								      Module pin
+//				MX6UL_PAD_SD1_CLK__GPIO2_IO17				/* 4  / SD1_CLK			*/
+//				MX6UL_PAD_SD1_CMD__GPIO2_IO16				/* 6  / SD1_CMD			*/
+//		 		MX6UL_PAD_GPIO1_IO00__GPIO1_IO00			/* 5  / USB_OTG1_ID 		*/
+//				MX6UL_PAD_SD1_DATA0__GPIO2_IO18				/* 8  / SD1_DATA0		*/
+//				MX6UL_PAD_SD1_DATA1__GPIO2_IO19				/* 10 / SD1_DATA1		*/
+//				MX6UL_PAD_SD1_DATA2__GPIO2_IO20				/* 12 / SD1_DATA2		*/
+//				MX6UL_PAD_UART1_RX_DATA__GPIO1_IO17			/* 13 / UART1_RX		*/
+//				MX6UL_PAD_SD1_DATA3__GPIO2_IO21				/* 14 / SD1_DATA3		*/
+//				MX6UL_PAD_UART1_TX_DATA__GPIO1_IO16			/* 15 / UART1_TX		*/ 
+//				MX6UL_PAD_UART1_RTS_B__GPIO1_IO19			/* 16 / SD1_CD			*/
+//				MX6UL_PAD_UART2_RX_DATA__GPIO1_IO21			/* 17 / UART2_RX		*/
+//				MX6UL_PAD_CSI_DATA04__GPIO4_IO25			/* 18 / ECSPI1_SCLK		*/
+//				MX6UL_PAD_UART2_TX_DATA__GPIO1_IO20			/* 19 / UART2_TX		*/
+//				MX6UL_PAD_CSI_DATA07__GPIO4_IO28			/* 20 / ECSPI1_MISO		*/
+//				MX6UL_PAD_UART3_RX_DATA__GPIO1_IO25			/* 21 / UART3_RX		*/
+//				MX6UL_PAD_CSI_DATA06__GPIO4_IO27			/* 22 / ECSPI1_MOSI		*/
+//				MX6UL_PAD_UART3_TX_DATA__GPIO1_IO24			/* 23 / UART3_TX		*/
+				MX6UL_PAD_CSI_DATA05__GPIO4_IO26	0x1b0b0		/* 24 / ECSPI1_SS0	SPI_CS0		*/ 
+//				MX6UL_PAD_CSI_MCLK__GPIO4_IO17		0x1b0b0		/* 25 / I2C1_SCL		*/
+//				MX6UL_PAD_LCD_DATA05__GPIO3_IO10			/* 26 / ECSPI1_SS1		*/
+//				MX6UL_PAD_CSI_PIXCLK__GPIO4_IO18	0x1b0b0		/* 27 / I2C1_SDA		*/
+//				MX6UL_PAD_UART4_RX_DATA__GPIO1_IO29			/* 28 / UART4_RX		*/
+//				MX6UL_PAD_UART4_TX_DATA__GPIO1_IO28			/* 30 / UART4_TX		*/
+//				MX6UL_PAD_UART3_RTS_B__GPIO1_IO27			/* 32 / UART3_RTS		*/
+//				MX6UL_PAD_CSI_DATA02__GPIO4_IO23			/* 34 / UART5_RTS		*/
+				MX6UL_PAD_CSI_DATA01__GPIO4_IO22			/* 36 / UART5_RX	LDC_DC	*/
+//				MX6UL_PAD_CSI_DATA00__GPIO4_IO21			/* 38 / UART5_TX		*/
+//				MX6UL_PAD_CSI_HSYNC__GPIO4_IO20				/* 40 / I2C2_SCL		*/
+//				MX6UL_PAD_CSI_VSYNC__GPIO4_IO19				/* 42 / I2C2_SCL		*/
+//				MX6UL_PAD_LCD_DATA09__GPIO3_IO14			/* 44 / CAN1_RX			*/
+//				MX6UL_PAD_LCD_DATA08__GPIO3_IO13			/* 46 / CAN1_TX			*/
+//				MX6UL_PAD_UART2_RTS_B__GPIO1_IO23			/* 48 / CAN2_RX			*/
+//				MX6UL_PAD_UART2_CTS_B__GPIO1_IO22			/* 50 / CAN2_TX			*/
+// FPC AUDIO Connector
+//				MX6UL_PAD_JTAG_TCK__GPIO1_IO14				/* 1  / SAI2_RXD		*/
+//				MX6UL_PAD_JTAG_TMS__GPIO1_IO11				/* 2  / SAI2_MCLK		*/
+//				MX6UL_PAD_JTAG_TDI__GPIO1_IO13				/* 3  / SAI2_BCLK		*/
+//				MX6UL_PAD_JTAG_TDO__GPIO1_IO12				/* 4  / SAI2_SYNC		*/
+//				MX6UL_PAD_JTAG_TRST_B__GPIO1_IO15			/* 5  / SAI2_TXD		*/
+//				MX6UL_PAD_JTAG_MOD__GPIO1_IO10				/* 6  / SPDIF_TX		*/
+//				MX6UL_PAD_LCD_DATA00__GPIO3_IO05			/* 7  / PWM1			*/
+//				MX6UL_PAD_LCD_DATA01__GPIO3_IO06			/* 8  / PWM2			*/
+//				MX6UL_PAD_SNVS_TAMPER9__GPIO5_IO09			/* 9  / TAMPER9			*/
+//				MX6UL_PAD_SNVS_TAMPER5__GPIO5_IO05			/* 10 / TAMPER5			*/
+//				MX6UL_PAD_LCD_DATA17__GPIO3_IO22			/* 11 / UART7_RX		*/
+//				MX6UL_PAD_LCD_DATA16__GPIO3_IO21			/* 12 / UART6_TX		*/
+
+// Full pin function list see in imx6ul-pinfunc.h and imx6ull-pinfunc.h 
+
+			>;
+		};
+
+
+
+		pinctrl_enet1: enet1grp {
+			fsl,pins = <
+				MX6UL_PAD_ENET1_RX_EN__ENET1_RX_EN	0x1b0b0
+				MX6UL_PAD_ENET1_RX_ER__ENET1_RX_ER	0x1b0b0
+				MX6UL_PAD_ENET1_RX_DATA0__ENET1_RDATA00	0x1b0b0
+				MX6UL_PAD_ENET1_RX_DATA1__ENET1_RDATA01	0x1b0b0
+				MX6UL_PAD_ENET1_TX_EN__ENET1_TX_EN	0x1b0b0
+				MX6UL_PAD_ENET1_TX_DATA0__ENET1_TDATA00	0x1b0b0
+				MX6UL_PAD_ENET1_TX_DATA1__ENET1_TDATA01	0x1b0b0
+				MX6UL_PAD_ENET1_TX_CLK__ENET1_REF_CLK1	0x4001b031
+			>;
+		};
+
+		pinctrl_enet2: enet2grp {
+			fsl,pins = <
+				MX6UL_PAD_GPIO1_IO07__ENET2_MDC		0x1b0b0
+				MX6UL_PAD_GPIO1_IO06__ENET2_MDIO	0x1b0b0
+				MX6UL_PAD_ENET2_RX_EN__ENET2_RX_EN	0x1b0b0
+				MX6UL_PAD_ENET2_RX_ER__ENET2_RX_ER	0x1b0b0
+				MX6UL_PAD_ENET2_RX_DATA0__ENET2_RDATA00	0x1b0b0
+				MX6UL_PAD_ENET2_RX_DATA1__ENET2_RDATA01	0x1b0b0
+				MX6UL_PAD_ENET2_TX_EN__ENET2_TX_EN	0x1b0b0
+				MX6UL_PAD_ENET2_TX_DATA0__ENET2_TDATA00	0x1b0b0
+				MX6UL_PAD_ENET2_TX_DATA1__ENET2_TDATA01	0x1b0b0
+				MX6UL_PAD_ENET2_TX_CLK__ENET2_REF_CLK2	0x4001b031
+			>;
+		};
+
+		pinctrl_flexcan1: flexcan1grp{
+			fsl,pins = <
+				MX6UL_PAD_LCD_DATA09__FLEXCAN1_RX	0x1b020  // check ok
+				MX6UL_PAD_LCD_DATA08__FLEXCAN1_TX	0x1b020
+			>;
+		};
+
+		pinctrl_flexcan2: flexcan2grp{
+			fsl,pins = <
+				MX6UL_PAD_UART2_RTS_B__FLEXCAN2_RX	0x1b020  // check ok
+				MX6UL_PAD_UART2_CTS_B__FLEXCAN2_TX	0x1b020
+			>;
+		};
+
+		pinctrl_i2c2: i2c2grp {
+			fsl,pins = <
+				MX6UL_PAD_CSI_HSYNC__I2C2_SCL 		0x4001b8b0 // check ok
+				MX6UL_PAD_CSI_VSYNC__I2C2_SDA 		0x4001b8b0
+			>;
+		};
+
+		pinctrl_ecspi1: spi1grp {
+			fsl,pins = <
+				MX6UL_PAD_CSI_DATA04__ECSPI1_SCLK	0x1b0b1
+				MX6UL_PAD_CSI_DATA06__ECSPI1_MOSI	0x1b0b1
+				MX6UL_PAD_CSI_DATA07__ECSPI1_MISO	0x1b0b1
+			>;
+		};
+
+		pinctrl_uart1: uart1grp {
+			fsl,pins = <
+				MX6UL_PAD_UART1_TX_DATA__UART1_DCE_TX 0x1b0b1		// check ok
+				MX6UL_PAD_UART1_RX_DATA__UART1_DCE_RX 0x1b0b15
+			>;
+		};
+
+		pinctrl_uart2: uart2grp {
+			fsl,pins = <
+				MX6UL_PAD_UART2_TX_DATA__UART2_DCE_TX	0x1b0b1		// check ok
+				MX6UL_PAD_UART2_RX_DATA__UART2_DCE_RX	0x1b0b1
+				MX6UL_PAD_UART2_CTS_B__UART2_DCE_CTS    0x1b0b1
+			>;
+		};
+
+		pinctrl_uart3: uart3grp {
+			fsl,pins = <
+				MX6UL_PAD_UART3_TX_DATA__UART3_DCE_TX	0x1b0b1		// check ok
+				MX6UL_PAD_UART3_RX_DATA__UART3_DCE_RX	0x1b0b1
+			>;
+		};
+
+		pinctrl_uart4: uart4grp {
+			fsl,pins = <
+				MX6UL_PAD_UART4_TX_DATA__UART4_DCE_TX	0x1b0b1		// check ok
+				MX6UL_PAD_UART4_RX_DATA__UART4_DCE_RX	0x1b0b1
+			>;
+		};
+
+		pinctrl_uart5: uart5grp {
+			fsl,pins = <
+				MX6UL_PAD_CSI_DATA00__UART5_DCE_TX	0x1b0b1		// check ok
+				MX6UL_PAD_CSI_DATA01__UART5_DCE_RX	0x1b0b1
+			>;
+		};
+
+		pinctrl_usdhc1: usdhc1grp {
+			fsl,pins = <
+				MX6UL_PAD_SD1_CMD__USDHC1_CMD     0x17059		// check ok
+				MX6UL_PAD_SD1_CLK__USDHC1_CLK     0x10071
+				MX6UL_PAD_SD1_DATA0__USDHC1_DATA0 0x17059
+				MX6UL_PAD_SD1_DATA1__USDHC1_DATA1 0x17059
+				MX6UL_PAD_SD1_DATA2__USDHC1_DATA2 0x17059
+				MX6UL_PAD_SD1_DATA3__USDHC1_DATA3 0x17059
+			>;
+		};
+
+		pinctrl_usdhc1_100mhz: usdhc1grp100mhz {
+			fsl,pins = <
+				MX6UL_PAD_SD1_CMD__USDHC1_CMD     0x170b9
+				MX6UL_PAD_SD1_CLK__USDHC1_CLK     0x100b9
+				MX6UL_PAD_SD1_DATA0__USDHC1_DATA0 0x170b9
+				MX6UL_PAD_SD1_DATA1__USDHC1_DATA1 0x170b9
+				MX6UL_PAD_SD1_DATA2__USDHC1_DATA2 0x170b9
+				MX6UL_PAD_SD1_DATA3__USDHC1_DATA3 0x170b9
+			>;
+		};
+
+		pinctrl_usdhc1_200mhz: usdhc1grp200mhz {
+			fsl,pins = <
+				MX6UL_PAD_SD1_CMD__USDHC1_CMD     0x170f9
+				MX6UL_PAD_SD1_CLK__USDHC1_CLK     0x100f9
+				MX6UL_PAD_SD1_DATA0__USDHC1_DATA0 0x170f9
+				MX6UL_PAD_SD1_DATA1__USDHC1_DATA1 0x170f9
+				MX6UL_PAD_SD1_DATA2__USDHC1_DATA2 0x170f9
+				MX6UL_PAD_SD1_DATA3__USDHC1_DATA3 0x170f9
+			>;
+		};
+
+		pinctrl_usdhc2: usdhc2grp {
+			fsl,pins = <
+				MX6UL_PAD_NAND_RE_B__USDHC2_CLK     0x10069
+				MX6UL_PAD_NAND_WE_B__USDHC2_CMD     0x17059
+				MX6UL_PAD_NAND_DATA00__USDHC2_DATA0 0x17059
+				MX6UL_PAD_NAND_DATA01__USDHC2_DATA1 0x17059
+				MX6UL_PAD_NAND_DATA02__USDHC2_DATA2 0x17059
+				MX6UL_PAD_NAND_DATA03__USDHC2_DATA3 0x17059
+				MX6UL_PAD_NAND_DATA04__USDHC2_DATA4 0x17059
+				MX6UL_PAD_NAND_DATA05__USDHC2_DATA5 0x17059
+				MX6UL_PAD_NAND_DATA06__USDHC2_DATA6 0x17059
+				MX6UL_PAD_NAND_DATA07__USDHC2_DATA7 0x17059
+
+			>;
+		};
+
+		pinctrl_wdog: wdoggrp {
+			fsl,pins = <
+				MX6UL_PAD_LCD_RESET__WDOG1_WDOG_ANY    0x30b0
+			>;
+		};
+	};
+};
+
+&usbotg1 {
+        dr_mode = "host";
+        srp-disable;
+        hnp-disable;
+        adp-disable;
+        vbus-supply = <&reg_usb1_vbus>;
+        status = "okay";
+};
+
+
+&usbotg2 {
+	dr_mode = "host";
+	disable-over-current;
+	vbus-supply = <&reg_usb2_vbus>;
+	status = "okay";
+};
+
+&usbphy1 {
+	tx-d-cal = <0x5>;
+};
+
+&usbphy2 {
+	tx-d-cal = <0x5>;
+};
+
+&usdhc1 {
+        pinctrl-names = "default", "state_100mhz", "state_200mhz";
+        pinctrl-0 = <&pinctrl_usdhc1>;
+        pinctrl-1 = <&pinctrl_usdhc1_100mhz>;
+        pinctrl-2 = <&pinctrl_usdhc1_200mhz>;
+        cd-gpios = <&gpio1 19 GPIO_ACTIVE_LOW>;
+        keep-power-in-suspend;
+        enable-sdio-wakeup;
+        vmmc-supply = <&reg_sd1_vmmc>;
+        status = "okay";
+};
+
+&usdhc2 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_usdhc2>;
+	bus-width = <8>;
+	no-1-8-v;
+	non-removable;
+	keep-power-in-suspend;
+	enable-sdio-wakeup;
+	status = "okay";
+};
+
+&wdog1 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_wdog>;
+	fsl,wdog_b;
+};
diff -urN a/arch/arm/boot/dts/Makefile b/arch/arm/boot/dts/Makefile
--- a/arch/arm/boot/dts/Makefile	2020-05-14 08:57:23.000000000 +0300
+++ b/arch/arm/boot/dts/Makefile	2020-07-03 19:08:07.850373021 +0300
@@ -558,6 +558,9 @@
 	imx6ul-tx6ul-0011.dtb \
 	imx6ul-tx6ul-mainboard.dtb \
 	imx6ull-14x14-evk.dtb \
+        o4-imx6ul-sodimm.dtb \
+        ev-imx6ul-m2.dtb \
+        ev-o4-imx6ul-nano.dtb \
 	imx6ull-colibri-eval-v3.dtb \
 	imx6ull-colibri-wifi-eval-v3.dtb
 dtb-$(CONFIG_SOC_IMX7D) += \
diff -urN a/arch/arm/boot/dts/o4-imx6ul-sodimm.dts b/arch/arm/boot/dts/o4-imx6ul-sodimm.dts
--- a/arch/arm/boot/dts/o4-imx6ul-sodimm.dts	1970-01-01 03:00:00.000000000 +0300
+++ b/arch/arm/boot/dts/o4-imx6ul-sodimm.dts	2020-07-03 20:44:08.725499042 +0300
@@ -0,0 +1,724 @@
+
+/*
+ * Copyright (C) 2015 Freescale Semiconductor, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ */
+
+/*
+  * O4-iMX6UL-SODIMM module
+  * www.out4.ru
+  * contact  dy@out4.ru
+  */
+
+
+/dts-v1/;
+
+#include <dt-bindings/input/input.h>
+#include "imx6ull.dtsi"
+
+/ {
+	model = "out4.ru o4-imx6ul-sodimm module";
+	compatible = "fsl,imx6ull-14x14-evk", "fsl,imx6ull";
+
+	chosen {
+		stdout-path = &uart1;
+	};
+
+	memory {
+		reg = <0x80000000 0x20000000>;
+	};
+
+	reserved-memory {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		ranges;
+
+		linux,cma {
+			compatible = "shared-dma-pool";
+			reusable;
+			size = <0x14000000>;
+			linux,cma-default;
+		};
+	};
+
+	backlight {
+		compatible = "pwm-backlight";
+		pwms = <&pwm1 0 5000000>;
+		brightness-levels = <0 4 8 16 32 64 128 255>;
+		default-brightness-level = <6>;
+		status = "okay";
+	};
+
+	pxp_v4l2 {
+		compatible = "fsl,imx6ul-pxp-v4l2", "fsl,imx6sx-pxp-v4l2", "fsl,imx6sl-pxp-v4l2";
+		status = "okay";
+	};
+
+	regulators {
+		compatible = "simple-bus";
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		reg_can_3v3: regulator@0 {
+			compatible = "regulator-fixed";
+			reg = <0>;
+			regulator-name = "can-3v3";
+			regulator-min-microvolt = <3300000>;
+			regulator-max-microvolt = <3300000>;
+			regulator-always-on;
+		};
+
+		reg_sd1_vmmc: regulator@1 {
+			compatible = "regulator-fixed";
+			reg = <1>;
+			regulator-name = "VSD_3V3";
+			regulator-min-microvolt = <3300000>;
+			regulator-max-microvolt = <3300000>;
+			regulator-always-on;
+		};
+
+		        reg_usb1_vbus: regulator@2 {
+                        compatible = "regulator-fixed";
+                        reg = <2>;
+                        regulator-name = "usb1_vbus";
+                        regulator-min-microvolt = <5000000>;
+                        regulator-max-microvolt = <5000000>;
+                        regulator-always-on;
+                };
+			reg_usb2_vbus: regulator@3 {
+                        compatible = "regulator-fixed";
+                        reg = <3>;
+                        regulator-name = "usb1_vbus";
+                        regulator-min-microvolt = <5000000>;
+                        regulator-max-microvolt = <5000000>;
+                        regulator-always-on;
+                };
+
+
+	};
+
+	spi4 {
+                compatible = "spi-gpio";
+                pinctrl-names = "default";
+                pinctrl-0 = <&pinctrl_spi4>;
+                pinctrl-assert-gpios = <&gpio5 8 GPIO_ACTIVE_LOW>;
+                status = "okay";
+                gpio-sck = <&gpio5 11 0>;
+                gpio-mosi = <&gpio5 10 0>;
+                cs-gpios = <&gpio5 7 0>;
+                num-chipselects = <1>;
+                #address-cells = <1>;
+                #size-cells = <0>;
+
+                gpio_spi: gpio_spi@0 {
+                        compatible = "fairchild,74hc595";
+                        gpio-controller;
+                        #gpio-cells = <2>;
+                        reg = <0>;
+                        registers-number = <1>;
+                        registers-default = /bits/ 8 <0x57>;
+                        spi-max-frequency = <100000>;
+                };
+	};
+};
+
+
+&fec1 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_enet1>;
+	phy-mode = "rmii";
+	phy-reset-gpios = <&gpio5 1 1>;
+	phy-reset-duration = <100>;
+	phy-handle = <&ethphy0>;
+	status = "okay";
+};
+
+&fec2 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_enet2>;
+	phy-mode = "rmii";
+	phy-handle = <&ethphy1>;
+	status = "okay";
+
+
+	mdio {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		ethphy0: ethernet-phy@0 {
+			compatible = "ethernet-phy-ieee802.3-c22";
+			reg = <0>;
+		};
+
+		ethphy1: ethernet-phy@1 {
+			compatible = "ethernet-phy-ieee802.3-c22";
+			reg = <1>;
+		};
+	};
+};
+
+&clks {
+        assigned-clocks = <&clks IMX6UL_CLK_PLL4_AUDIO_DIV>;
+        assigned-clock-rates = <786432000>;
+};
+
+
+&can1 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_flexcan1>;
+	xceiver-supply = <&reg_can_3v3>;
+	status = "okay";
+};
+
+&can2 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_flexcan2>;
+	xceiver-supply = <&reg_can_3v3>;
+	status = "okay";
+};
+
+&gpc {
+	fsl,cpu_pupscr_sw2iso = <0x1>;
+	fsl,cpu_pupscr_sw = <0x0>;
+	fsl,cpu_pdnscr_iso2sw = <0x1>;
+	fsl,cpu_pdnscr_iso = <0x1>;
+	fsl,ldo-bypass = <0>; /* DCDC, ldo-enable */
+};
+
+
+&i2c2 {
+	clock_frequency = <100000>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_i2c2>;
+	status = "okay";
+
+	codec: wm8960@1a {
+		compatible = "wlf,wm8960";
+		reg = <0x1a>;
+		clocks = <&clks IMX6UL_CLK_SAI2>;
+		clock-names = "mclk";
+		wlf,shared-lrclk;
+	};
+};
+
+&iomuxc_snvs {
+        pinctrl-names = "default_snvs";
+        pinctrl-0 = <&pinctrl_hog_2>;
+        imx6ul-evk {
+                pinctrl_hog_2: hoggrp-2 {
+                        fsl,pins = <
+                                /* ETH_RES */
+                                MX6ULL_PAD_SNVS_TAMPER1__GPIO5_IO01      0x80000000
+                        >;
+                };
+		 pinctrl_sai2_hp_det_b: sai2_hp_det_grp {
+                        fsl,pins = <
+                                MX6ULL_PAD_SNVS_TAMPER4__GPIO5_IO04   0x17059
+                        >;
+                };
+
+
+        };
+};
+
+
+&lcdif {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_lcdif_dat
+		     &pinctrl_lcdif_ctrl>;
+	display = <&display0>;
+	status = "okay";
+
+	display0: display {
+		bits-per-pixel = <16>;
+		bus-width = <24>;
+
+		display-timings {
+			native-mode = <&timing0>;
+			timing0: timing0 {
+			clock-frequency = <35000000>;
+			hactive = <800>;
+			vactive = <480>;
+			hfront-porch = <40>;
+			hback-porch = <40>;
+			hsync-len = <48>;
+			vback-porch = <29>;
+			vfront-porch = <13>;
+			vsync-len = <3>;
+
+			hsync-active = <0>;
+			vsync-active = <0>;
+			de-active = <1>;
+			pixelclk-active = <0>;
+			};
+		};
+	};
+};
+
+&pwm1 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_pwm1>;
+	status = "okay";
+};
+
+
+&sai2 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_sai2
+		    &pinctrl_sai2_hp_det_b>;
+
+	assigned-clocks = <&clks IMX6UL_CLK_SAI2_SEL>,
+			  <&clks IMX6UL_CLK_SAI2>;
+	assigned-clock-parents = <&clks IMX6UL_CLK_PLL4_AUDIO_DIV>;
+	assigned-clock-rates = <0>, <12288000>;
+	status = "okay";
+};
+
+
+&tsc {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_tsc>;
+	xnur-gpio = <&gpio1 3 GPIO_ACTIVE_LOW>;
+	measure-delay-time = <0xffff>;
+	pre-charge-time = <0xfff>;
+	status = "okay";
+};
+
+&uart1 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_uart1>;
+	status = "okay";
+};
+
+&uart2 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_uart2>;
+	/* fsl,uart-has-rtscts; */ 
+	/* for DTE mode, add below change */
+	/* fsl,dte-mode; */
+	status = "okay";
+};
+
+&uart3 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_uart3>;
+	/* fsl,uart-has-rtscts; */ 
+	/* for DTE mode, add below change */
+	/* fsl,dte-mode; */
+	status = "okay";
+};
+
+&uart4 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_uart4>;
+	status = "okay";
+};
+
+//&uart6 {
+//	pinctrl-names = "default";
+//	pinctrl-0 = <&pinctrl_uart6>;  /* ENET2 mux */
+//	status = "okay";
+//};
+
+
+&qspi {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_qspi>;
+	status = "okay";
+	ddrsmp=<0>;
+
+	flash0: n25q256a@0 {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "micron,n25q256a";
+		spi-max-frequency = <29000000>;
+		spi-nor,ddr-quad-read-dummy = <6>;
+		reg = <0>;
+	};
+}; 
+
+
+&usbotg1 {
+	pinctrl-names = "default";
+        pinctrl-0 = <&pinctrl_usb_otg1_id>;
+	dr_mode = "otg";
+	srp-disable;
+	hnp-disable;
+	adp-disable;
+	vbus-supply = <&reg_usb1_vbus>;
+	status = "okay";
+};
+
+&usbotg2 {
+	dr_mode = "host";
+	disable-over-current;
+	vbus-supply = <&reg_usb2_vbus>;
+	status = "okay";
+};
+
+&usbphy1 {
+	tx-d-cal = <0x5>;
+};
+
+&usbphy2 {
+	tx-d-cal = <0x5>;
+};
+
+&usdhc1 {
+	pinctrl-names = "default", "state_100mhz", "state_200mhz";
+	pinctrl-0 = <&pinctrl_usdhc1>;
+	pinctrl-1 = <&pinctrl_usdhc1_100mhz>;
+	pinctrl-2 = <&pinctrl_usdhc1_200mhz>;
+	cd-gpios = <&gpio1 19 GPIO_ACTIVE_LOW>;
+	keep-power-in-suspend;
+	enable-sdio-wakeup;
+	vmmc-supply = <&reg_sd1_vmmc>;
+	status = "okay";
+};
+
+&usdhc2 {
+	pinctrl-names = "default", "state_100mhz", "state_200mhz";
+	pinctrl-0 = <&pinctrl_usdhc2_8bit>;
+	pinctrl-1 = <&pinctrl_usdhc2_8bit_100mhz>;
+	pinctrl-2 = <&pinctrl_usdhc2_8bit_200mhz>;
+	bus-width = <8>; 
+	non-removable;
+	status = "okay";
+};
+
+&wdog1 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_wdog>;
+	fsl,wdog_b;
+};
+
+
+&iomuxc {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_hog_1>;
+	imx6ul-evk {
+		pinctrl_hog_1: hoggrp-1 {
+			fsl,pins = <
+				MX6UL_PAD_UART1_RTS_B__GPIO1_IO19	0x17059 /* SD1 CD */
+				MX6UL_PAD_GPIO1_IO05__USDHC1_VSELECT	0x17059 /* SD1 VSELECT */
+				MX6UL_PAD_GPIO1_IO09__GPIO1_IO09        0x17059 /* SD1 RESET */
+				MX6UL_PAD_SNVS_TAMPER0__GPIO5_IO00	0x80000000
+			>;
+		};
+
+		pinctrl_csi1: csi1grp {
+			fsl,pins = <
+				MX6UL_PAD_CSI_MCLK__CSI_MCLK		0x1b088
+				MX6UL_PAD_CSI_PIXCLK__CSI_PIXCLK	0x1b088
+				MX6UL_PAD_CSI_VSYNC__CSI_VSYNC		0x1b088
+				MX6UL_PAD_CSI_HSYNC__CSI_HSYNC		0x1b088
+				MX6UL_PAD_CSI_DATA00__CSI_DATA02	0x1b088
+				MX6UL_PAD_CSI_DATA01__CSI_DATA03	0x1b088
+				MX6UL_PAD_CSI_DATA02__CSI_DATA04	0x1b088
+				MX6UL_PAD_CSI_DATA03__CSI_DATA05	0x1b088
+				MX6UL_PAD_CSI_DATA04__CSI_DATA06	0x1b088
+				MX6UL_PAD_CSI_DATA05__CSI_DATA07	0x1b088
+				MX6UL_PAD_CSI_DATA06__CSI_DATA08	0x1b088
+				MX6UL_PAD_CSI_DATA07__CSI_DATA09	0x1b088
+			>;
+		};
+
+		pinctrl_enet1: enet1grp {
+			fsl,pins = <
+				MX6UL_PAD_ENET1_RX_EN__ENET1_RX_EN	0x1b0b0
+				MX6UL_PAD_ENET1_RX_ER__ENET1_RX_ER	0x1b0b0
+				MX6UL_PAD_ENET1_RX_DATA0__ENET1_RDATA00	0x1b0b0
+				MX6UL_PAD_ENET1_RX_DATA1__ENET1_RDATA01	0x1b0b0
+				MX6UL_PAD_ENET1_TX_EN__ENET1_TX_EN	0x1b0b0
+				MX6UL_PAD_ENET1_TX_DATA0__ENET1_TDATA00	0x1b0b0
+				MX6UL_PAD_ENET1_TX_DATA1__ENET1_TDATA01	0x1b0b0
+				MX6UL_PAD_ENET1_TX_CLK__ENET1_REF_CLK1	0x4001b031
+			>;
+		};
+
+		pinctrl_enet2: enet2grp {
+			fsl,pins = <
+				MX6UL_PAD_GPIO1_IO07__ENET2_MDC		0x1b0b0
+				MX6UL_PAD_GPIO1_IO06__ENET2_MDIO	0x1b0b0
+				MX6UL_PAD_ENET2_RX_EN__ENET2_RX_EN	0x1b0b0
+				MX6UL_PAD_ENET2_RX_ER__ENET2_RX_ER	0x1b0b0
+				MX6UL_PAD_ENET2_RX_DATA0__ENET2_RDATA00	0x1b0b0
+				MX6UL_PAD_ENET2_RX_DATA1__ENET2_RDATA01	0x1b0b0
+				MX6UL_PAD_ENET2_TX_EN__ENET2_TX_EN	0x1b0b0
+				MX6UL_PAD_ENET2_TX_DATA0__ENET2_TDATA00	0x1b0b0
+				MX6UL_PAD_ENET2_TX_DATA1__ENET2_TDATA01	0x1b0b0
+				MX6UL_PAD_ENET2_TX_CLK__ENET2_REF_CLK2	0x4001b031
+			>;
+		};
+
+		pinctrl_flexcan1: flexcan1grp{
+			fsl,pins = <
+				MX6UL_PAD_UART3_RTS_B__FLEXCAN1_RX	0x1b020
+				MX6UL_PAD_UART3_CTS_B__FLEXCAN1_TX	0x1b020
+			>;
+		};
+
+		pinctrl_flexcan2: flexcan2grp{
+			fsl,pins = <
+				MX6UL_PAD_UART2_RTS_B__FLEXCAN2_RX	0x1b020
+				MX6UL_PAD_UART2_CTS_B__FLEXCAN2_TX	0x1b020
+			>;
+		};
+
+		pinctrl_i2c1: i2c1grp {
+			fsl,pins = <
+				MX6UL_PAD_UART4_TX_DATA__I2C1_SCL 0x4001b8b0
+				MX6UL_PAD_UART4_RX_DATA__I2C1_SDA 0x4001b8b0
+			>;
+		};
+
+		pinctrl_i2c2: i2c2grp {
+			fsl,pins = <
+				MX6UL_PAD_UART5_TX_DATA__I2C2_SCL 0x4001b8b0
+				MX6UL_PAD_UART5_RX_DATA__I2C2_SDA 0x4001b8b0
+			>;
+		};
+
+		pinctrl_lcdif_dat: lcdifdatgrp {
+			fsl,pins = <
+				MX6UL_PAD_LCD_DATA00__LCDIF_DATA00  0x79
+				MX6UL_PAD_LCD_DATA01__LCDIF_DATA01  0x79
+				MX6UL_PAD_LCD_DATA02__LCDIF_DATA02  0x79
+				MX6UL_PAD_LCD_DATA03__LCDIF_DATA03  0x79
+				MX6UL_PAD_LCD_DATA04__LCDIF_DATA04  0x79
+				MX6UL_PAD_LCD_DATA05__LCDIF_DATA05  0x79
+				MX6UL_PAD_LCD_DATA06__LCDIF_DATA06  0x79
+				MX6UL_PAD_LCD_DATA07__LCDIF_DATA07  0x79
+				MX6UL_PAD_LCD_DATA08__LCDIF_DATA08  0x79
+				MX6UL_PAD_LCD_DATA09__LCDIF_DATA09  0x79
+				MX6UL_PAD_LCD_DATA10__LCDIF_DATA10  0x79
+				MX6UL_PAD_LCD_DATA11__LCDIF_DATA11  0x79
+				MX6UL_PAD_LCD_DATA12__LCDIF_DATA12  0x79
+				MX6UL_PAD_LCD_DATA13__LCDIF_DATA13  0x79
+				MX6UL_PAD_LCD_DATA14__LCDIF_DATA14  0x79
+				MX6UL_PAD_LCD_DATA15__LCDIF_DATA15  0x79
+				MX6UL_PAD_LCD_DATA16__LCDIF_DATA16  0x79
+				MX6UL_PAD_LCD_DATA17__LCDIF_DATA17  0x79
+				MX6UL_PAD_LCD_DATA18__LCDIF_DATA18  0x79
+				MX6UL_PAD_LCD_DATA19__LCDIF_DATA19  0x79
+				MX6UL_PAD_LCD_DATA20__LCDIF_DATA20  0x79
+				MX6UL_PAD_LCD_DATA21__LCDIF_DATA21  0x79
+				MX6UL_PAD_LCD_DATA22__LCDIF_DATA22  0x79
+				MX6UL_PAD_LCD_DATA23__LCDIF_DATA23  0x79
+			>;
+		};
+
+		pinctrl_lcdif_ctrl: lcdifctrlgrp {
+			fsl,pins = <
+				MX6UL_PAD_LCD_CLK__LCDIF_CLK	    0x79
+				MX6UL_PAD_LCD_ENABLE__LCDIF_ENABLE  0x79
+				MX6UL_PAD_LCD_HSYNC__LCDIF_HSYNC    0x79
+				MX6UL_PAD_LCD_VSYNC__LCDIF_VSYNC    0x79
+			>;
+		};
+
+		pinctrl_pwm1: pwm1grp {
+			fsl,pins = <
+				MX6UL_PAD_GPIO1_IO08__PWM1_OUT   0x110b0
+			>;
+		};
+
+		pinctrl_qspi: qspigrp {
+			fsl,pins = <
+				MX6UL_PAD_NAND_WP_B__QSPI_A_SCLK      0x70a1
+				MX6UL_PAD_NAND_READY_B__QSPI_A_DATA00 0x70a1
+				MX6UL_PAD_NAND_CE0_B__QSPI_A_DATA01   0x70a1
+				MX6UL_PAD_NAND_CE1_B__QSPI_A_DATA02   0x70a1
+				MX6UL_PAD_NAND_CLE__QSPI_A_DATA03     0x70a1
+				MX6UL_PAD_NAND_DQS__QSPI_A_SS0_B      0x70a1
+			>;
+		};
+
+		pinctrl_sai2: sai2grp {
+			fsl,pins = <
+				MX6UL_PAD_JTAG_TDI__SAI2_TX_BCLK	0x17088
+				MX6UL_PAD_JTAG_TDO__SAI2_TX_SYNC	0x17088
+				MX6UL_PAD_JTAG_TRST_B__SAI2_TX_DATA	0x11088
+				MX6UL_PAD_JTAG_TCK__SAI2_RX_DATA	0x11088
+				MX6UL_PAD_JTAG_TMS__SAI2_MCLK		0x17088
+			>;
+		};
+
+		pinctrl_sim2_1: sim2grp-1 {
+			fsl,pins = <
+				MX6UL_PAD_CSI_DATA03__SIM2_PORT1_PD		0xb808
+				MX6UL_PAD_CSI_DATA04__SIM2_PORT1_CLK		0x11
+				MX6UL_PAD_CSI_DATA05__SIM2_PORT1_RST_B		0xb810
+				MX6UL_PAD_CSI_DATA06__SIM2_PORT1_SVEN		0xb810
+				MX6UL_PAD_CSI_DATA07__SIM2_PORT1_TRXD		0xb811
+				MX6UL_PAD_CSI_DATA02__GPIO4_IO23		0x3008
+			>;
+		};
+
+		pinctrl_spi4: spi4grp {
+			fsl,pins = <
+				MX6UL_PAD_BOOT_MODE0__GPIO5_IO10	0x70a1
+				MX6UL_PAD_BOOT_MODE1__GPIO5_IO11	0x70a1
+				MX6UL_PAD_SNVS_TAMPER7__GPIO5_IO07	0x70a1
+				MX6UL_PAD_SNVS_TAMPER8__GPIO5_IO08	0x80000000
+			>;
+		};
+
+		pinctrl_tsc: tscgrp {
+			fsl,pins = <
+				MX6UL_PAD_GPIO1_IO01__GPIO1_IO01	0xb0
+				MX6UL_PAD_GPIO1_IO02__GPIO1_IO02	0xb0
+				MX6UL_PAD_GPIO1_IO03__GPIO1_IO03	0xb0
+				MX6UL_PAD_GPIO1_IO04__GPIO1_IO04	0xb0
+			>;
+		};
+
+		pinctrl_uart1: uart1grp {
+			fsl,pins = <
+				MX6UL_PAD_UART1_TX_DATA__UART1_DCE_TX 0x1b0b1
+				MX6UL_PAD_UART1_RX_DATA__UART1_DCE_RX 0x1b0b1
+			>;
+		};
+
+		pinctrl_uart2: uart2grp {
+			fsl,pins = <
+				MX6UL_PAD_UART2_TX_DATA__UART2_DCE_TX	0x1b0b1
+				MX6UL_PAD_UART2_RX_DATA__UART2_DCE_RX	0x1b0b1
+		//		MX6UL_PAD_UART2_RX_DATA__UART2_DCE_RTS	0x1b0b1   /* CAN2 mux */
+		//		MX6UL_PAD_UART2_TX_DATA__UART2_DCE_CTS	0x1b0b1	  /* CAN2 mux */ 	
+			>;
+		};
+
+		pinctrl_uart3: uart3grp {
+			fsl,pins = <
+				MX6UL_PAD_UART3_TX_DATA__UART3_DCE_TX	0x1b0b1
+				MX6UL_PAD_UART3_RX_DATA__UART3_DCE_RX	0x1b0b1
+		//		MX6UL_PAD_UART3_RTS_B__UART3_DCE_RTS	0x1b0b1   /* CAN1 mux */
+		//		MX6UL_PAD_UART3_CTS_B__UART3_DCE_CTS	0x1b0b1	  /* CAN1 mux */ 	
+			>;
+		};
+
+		pinctrl_uart4: uart4grp {
+			fsl,pins = <
+				MX6UL_PAD_UART4_TX_DATA__UART4_DCE_TX	0x1b0b1
+				MX6UL_PAD_UART4_RX_DATA__UART4_DCE_RX	0x1b0b1
+			>;
+		};
+
+		pinctrl_uart6: uart6grp {
+			fsl,pins = <
+				MX6UL_PAD_ENET2_RX_DATA0__UART6_DCE_TX	0x1b0b1   /* ENET2 mux */
+				MX6UL_PAD_ENET2_RX_DATA1__UART6_DCE_RX	0x1b0b1   /* ENET2 mux */
+			>;
+		};
+		pinctrl_usb_otg1_id: usbotg1idgrp {
+                fsl,pins = <
+                    		MX6UL_PAD_GPIO1_IO00__ANATOP_OTG1_ID    0x17059
+            	    >;
+    		};
+
+
+
+
+		pinctrl_usdhc1: usdhc1grp {
+			fsl,pins = <
+				MX6UL_PAD_SD1_CMD__USDHC1_CMD     0x17059
+				MX6UL_PAD_SD1_CLK__USDHC1_CLK     0x10071
+				MX6UL_PAD_SD1_DATA0__USDHC1_DATA0 0x17059
+				MX6UL_PAD_SD1_DATA1__USDHC1_DATA1 0x17059
+				MX6UL_PAD_SD1_DATA2__USDHC1_DATA2 0x17059
+				MX6UL_PAD_SD1_DATA3__USDHC1_DATA3 0x17059
+			>;
+		};
+
+		pinctrl_usdhc1_100mhz: usdhc1grp100mhz {
+			fsl,pins = <
+				MX6UL_PAD_SD1_CMD__USDHC1_CMD     0x170b9
+				MX6UL_PAD_SD1_CLK__USDHC1_CLK     0x100b9
+				MX6UL_PAD_SD1_DATA0__USDHC1_DATA0 0x170b9
+				MX6UL_PAD_SD1_DATA1__USDHC1_DATA1 0x170b9
+				MX6UL_PAD_SD1_DATA2__USDHC1_DATA2 0x170b9
+				MX6UL_PAD_SD1_DATA3__USDHC1_DATA3 0x170b9
+			>;
+		};
+
+		pinctrl_usdhc1_200mhz: usdhc1grp200mhz {
+			fsl,pins = <
+				MX6UL_PAD_SD1_CMD__USDHC1_CMD     0x170f9
+				MX6UL_PAD_SD1_CLK__USDHC1_CLK     0x100f9
+				MX6UL_PAD_SD1_DATA0__USDHC1_DATA0 0x170f9
+				MX6UL_PAD_SD1_DATA1__USDHC1_DATA1 0x170f9
+				MX6UL_PAD_SD1_DATA2__USDHC1_DATA2 0x170f9
+				MX6UL_PAD_SD1_DATA3__USDHC1_DATA3 0x170f9
+			>;
+		};
+
+		pinctrl_usdhc2: usdhc2grp {
+			fsl,pins = <
+				MX6UL_PAD_NAND_RE_B__USDHC2_CLK     0x10069
+				MX6UL_PAD_NAND_WE_B__USDHC2_CMD     0x17059
+				MX6UL_PAD_NAND_DATA00__USDHC2_DATA0 0x17059
+				MX6UL_PAD_NAND_DATA01__USDHC2_DATA1 0x17059
+				MX6UL_PAD_NAND_DATA02__USDHC2_DATA2 0x17059
+				MX6UL_PAD_NAND_DATA03__USDHC2_DATA3 0x17059
+			>;
+		};
+
+		pinctrl_usdhc2_8bit: usdhc2grp_8bit {
+			fsl,pins = <
+				MX6UL_PAD_NAND_RE_B__USDHC2_CLK     0x10069
+				MX6UL_PAD_NAND_WE_B__USDHC2_CMD     0x17059
+				MX6UL_PAD_NAND_DATA00__USDHC2_DATA0 0x17059
+				MX6UL_PAD_NAND_DATA01__USDHC2_DATA1 0x17059
+				MX6UL_PAD_NAND_DATA02__USDHC2_DATA2 0x17059
+				MX6UL_PAD_NAND_DATA03__USDHC2_DATA3 0x17059
+				MX6UL_PAD_NAND_DATA04__USDHC2_DATA4 0x17059
+				MX6UL_PAD_NAND_DATA05__USDHC2_DATA5 0x17059
+				MX6UL_PAD_NAND_DATA06__USDHC2_DATA6 0x17059
+				MX6UL_PAD_NAND_DATA07__USDHC2_DATA7 0x17059
+			>;
+		};
+
+		pinctrl_usdhc2_8bit_100mhz: usdhc2grp_8bit_100mhz {
+			fsl,pins = <
+				MX6UL_PAD_NAND_RE_B__USDHC2_CLK     0x100b9
+				MX6UL_PAD_NAND_WE_B__USDHC2_CMD     0x170b9
+				MX6UL_PAD_NAND_DATA00__USDHC2_DATA0 0x170b9
+				MX6UL_PAD_NAND_DATA01__USDHC2_DATA1 0x170b9
+				MX6UL_PAD_NAND_DATA02__USDHC2_DATA2 0x170b9
+				MX6UL_PAD_NAND_DATA03__USDHC2_DATA3 0x170b9
+				MX6UL_PAD_NAND_DATA04__USDHC2_DATA4 0x170b9
+				MX6UL_PAD_NAND_DATA05__USDHC2_DATA5 0x170b9
+				MX6UL_PAD_NAND_DATA06__USDHC2_DATA6 0x170b9
+				MX6UL_PAD_NAND_DATA07__USDHC2_DATA7 0x170b9
+			>;
+		};
+
+		pinctrl_usdhc2_8bit_200mhz: usdhc2grp_8bit_200mhz {
+			fsl,pins = <
+				MX6UL_PAD_NAND_RE_B__USDHC2_CLK     0x100f9
+				MX6UL_PAD_NAND_WE_B__USDHC2_CMD     0x170f9
+				MX6UL_PAD_NAND_DATA00__USDHC2_DATA0 0x170f9
+				MX6UL_PAD_NAND_DATA01__USDHC2_DATA1 0x170f9
+				MX6UL_PAD_NAND_DATA02__USDHC2_DATA2 0x170f9
+				MX6UL_PAD_NAND_DATA03__USDHC2_DATA3 0x170f9
+				MX6UL_PAD_NAND_DATA04__USDHC2_DATA4 0x170f9
+				MX6UL_PAD_NAND_DATA05__USDHC2_DATA5 0x170f9
+				MX6UL_PAD_NAND_DATA06__USDHC2_DATA6 0x170f9
+				MX6UL_PAD_NAND_DATA07__USDHC2_DATA7 0x170f9
+			>;
+		};
+		pinctrl_wdog: wdoggrp {
+			fsl,pins = <
+				MX6UL_PAD_LCD_RESET__WDOG1_WDOG_ANY    0x30b0
+			>;
+		};
+	};
+};
+
+
+
+
